File: gfg-solutions/Median of two sorted arrays/Median of two sorted arrays.java
class Solution {
    public int sumOfMiddleElements(int[] arr1, int[] arr2) {
        int n = arr1.length + arr2.length;

        return kthElement(n / 2, arr1, arr2) + kthElement((n / 2) + 1, arr1, arr2);
    }

    private int kthElement(int k, int[] arr1, int[] arr2) {
        int n = arr1.length;
        int m = arr2.length;

        if (n < m)
            return kthElement(k, arr2, arr1);

        int low = Math.max(0, k - m), high = Math.min(n, k);

        while (low <= high) {
            int cut1 = (low + high) / 2;
            int cut2 = k - cut1;

            int l1 = cut1 == 0 ? Integer.MIN_VALUE : arr1[cut1 - 1];
            int l2 = cut2 == 0 ? Integer.MIN_VALUE : arr2[cut2 - 1];

            int r1 = cut1 == n ? Integer.MAX_VALUE : arr1[cut1];
            int r2 = cut2 == m ? Integer.MAX_VALUE : arr2[cut2];

            if (l1 <= r1 && l2 <= r2)
                return Math.max(l1, l2);
            else if (l1 > l2)
                high = cut1 - 1;
            else
                low = cut1 + 1;
        }

        return 0;
    }
}

File: gfg-solutions/Median of two sorted arrays/Median of two sorted arrays.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    int SumofMiddleElements(vector<int> &arr1, vector<int> &arr2)
    {
        int n = arr1.size() + arr2.size();

        int val1 = kthElement(n / 2, arr1, arr2);
        int val2 = kthElement((n / 2) + 1, arr1, arr2);
        return val1 + val2;
    }

    int kthElement(int k, vector<int> &arr1, vector<int> &arr2)
    {
        int n = arr1.size(), m = arr2.size();
        int low = max(0, k - m), high = min(n, k);
        while (low <= high)
        {
            int pickedFirst = low + (high - low) / 2;
            int pickedSecond = k - pickedFirst;

            int pickedFirstElement = (pickedFirst - 1 >= 0) ? arr1[pickedFirst - 1] : -1;
            int pickedFirstNext = (pickedFirst < n) ? arr1[pickedFirst] : INT_MAX;

            int pickedSecondElement = (pickedSecond - 1 >= 0) ? arr2[pickedSecond - 1] : -1;
            int pickedSecondNext = (pickedSecond < m) ? arr2[pickedSecond] : INT_MAX;

            if (pickedFirstElement <= pickedSecondNext && pickedSecondElement <= pickedFirstNext)
            {
                // the largest one on the left side will be the kth element
                return max(pickedFirstElement, pickedSecondElement);
            }
            else if (pickedFirstElement > pickedSecondNext)
            {
                // this means that the pickedFirst should move back
                // as we have scope to increase the pickedSecond
                high = pickedFirst - 1;
            }
            else
            {
                low = pickedFirst + 1;
            }
        }

        return -1;
    }
};